// @openshift-console/dynamic-plugin-sdk-webpack/lib/extensions/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuOptionType = exports.isActionFilter = exports.isActionGroup = exports.isResourceActionProvider = exports.isActionProvider = void 0;
// Type Guards
const isActionProvider = (e) => {
    return e.type === 'console.action/provider';
};
exports.isActionProvider = isActionProvider;
const isResourceActionProvider = (e) => {
    return e.type === 'console.action/resource-provider';
};
exports.isResourceActionProvider = isResourceActionProvider;
const isActionGroup = (e) => {
    return e.type === 'console.action/group';
};
exports.isActionGroup = isActionGroup;
const isActionFilter = (e) => {
    return e.type === 'console.action/filter';
};
exports.isActionFilter = isActionFilter;
var MenuOptionType;
(function (MenuOptionType) {
    MenuOptionType[MenuOptionType["GROUP_MENU"] = 0] = "GROUP_MENU";
    MenuOptionType[MenuOptionType["SUB_MENU"] = 1] = "SUB_MENU";
    MenuOptionType[MenuOptionType["ATOMIC_MENU"] = 2] = "ATOMIC_MENU";
})(MenuOptionType = exports.MenuOptionType || (exports.MenuOptionType = {}));
